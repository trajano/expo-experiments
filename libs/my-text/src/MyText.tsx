import { createContext, forwardRef, useContext, useMemo } from 'react';
import { StyleSheet, Text, type TextProps, type TextStyle } from 'react-native';
import { mapToEmbeddedFontFamily, mapToGoogleFontKey } from './mapToGoogleFontKey';
import type { FontFamilyMapper, LineHeightMapper } from './MyText.types';

/**
 * FontStyleContext is used to define the shape of the font styles
 * that will be tracked and inherited by nested MyText components.
 */
interface FontStyleContext {
    fontFamily?: string;
    fontStyle?: TextStyle['fontStyle'];
    fontWeight?: TextStyle['fontWeight'];
    mapToFontFamily: FontFamilyMapper;
    mapToLineHeight: LineHeightMapper;
}

/**
 * MyTextContext is a React Context that provides the font styles to
 * all nested MyText components.
 */
const MyTextContext = createContext<FontStyleContext>({
    mapToFontFamily: mapToGoogleFontKey,
    mapToLineHeight: () => undefined,
});

export type MyTextProps = TextProps & {
    /**
     * An optional mapper that would get the final font family given the logical font family, font weight and font style.
     * This defalts to {@link mapToGoogleFontKey}
     */
    mapToFontFamily?: FontFamilyMapper,
    /**
     * An optional mapper that would take the fontSize and translate it to a lineHeight value to ensure consistency across platforms.
     * This defalts to always return `undefined` which lets React Native handle it natively.
     */
    mapToLineHeight?: LineHeightMapper,
}
/**
 * MyText is a custom Text component that inherits font styles
 * (fontFamily, fontStyle, fontWeight) from its parent MyText components
 * unless explicitly overridden.  It presumes the use of Expo Google Fonts  https://docs.expo.dev/develop/user-interface/fonts/#use-google-fonts 
 * and are loaded using useExpoGoogleFonts hook
 *
 * @param props - Props passed to the Text component.
 * @param ref - Ref forwarded to the Text component.
 * @returns A Text component with inherited font styles.
 */
export const MyText = forwardRef<Text, MyTextProps>(({ children, mapToFontFamily, mapToLineHeight, style, ...props }, ref) => {
    // Retrieve the parent styles from context
    const parentStyles = useContext(MyTextContext);

    // Flatten the incoming style to simplify processing
    const flattenedStyle = useMemo(() => StyleSheet.flatten(style) ?? {}, [style]);

    // Use a single useMemo to compute both currentFontStyles and combinedStyle
    const { currentFontStyles, combinedStyle } = useMemo(() => {
        const mapperToFontFamily = mapToFontFamily ?? mapToGoogleFontKey;
        const mapperToLineHeight = mapToLineHeight ?? (() => undefined);
        const currentFontStyles: FontStyleContext = {
            fontFamily: flattenedStyle.fontFamily ?? parentStyles.fontFamily,
            fontStyle: flattenedStyle.fontStyle ?? parentStyles.fontStyle,
            fontWeight: flattenedStyle.fontWeight ?? parentStyles.fontWeight,
            mapToFontFamily: mapperToFontFamily,
            mapToLineHeight: mapperToLineHeight
        };

        const mappedFontFamily = mapperToFontFamily(
            currentFontStyles.fontFamily,
            currentFontStyles.fontWeight,
            currentFontStyles.fontStyle,
        );

        const mappedLineHeight = flattenedStyle.lineHeight ?? mapperToLineHeight(flattenedStyle.fontSize);

        const combinedStyle: TextStyle = {
            ...flattenedStyle,
            fontFamily: mappedFontFamily ?? currentFontStyles.fontFamily, // Fallback to original fontFamily if mapping fails
            fontStyle: mappedFontFamily ? undefined : currentFontStyles.fontStyle, // Fallback to original fontStyle if mapping fails
            fontWeight: mappedFontFamily ? undefined : currentFontStyles.fontWeight, // Fallback to original fontWeight if mapping fails
            lineHeight: mappedLineHeight,
        };

        return { currentFontStyles, combinedStyle };
    }, [flattenedStyle, parentStyles]);

    return (
        <MyTextContext.Provider value={currentFontStyles}>
            <Text ref={ref} style={combinedStyle} {...props}>
                {children}
            </Text>
        </MyTextContext.Provider>
    );
});

/**
 * MyText is a custom Text component that inherits font styles
 * (fontFamily, fontStyle, fontWeight) from its parent MyText components
 * unless explicitly overridden.  It presumes the use of Expo Google Fonts  https://docs.expo.dev/develop/user-interface/fonts/#use-google-fonts 
 * and are loaded using useExpoGoogleFonts hook
 *
 * @param props - Props passed to the Text component.
 * @param ref - Ref forwarded to the Text component.
 * @returns A Text component with inherited font styles.
 */
export const MyTextE = forwardRef<Text, MyTextProps>(({ children, mapToFontFamily, mapToLineHeight, style, ...props }, ref) => {
    // Retrieve the parent styles from context
    const parentStyles = useContext(MyTextContext);

    // Flatten the incoming style to simplify processing
    const flattenedStyle = useMemo(() => StyleSheet.flatten(style) ?? {}, [style]);

    // Use a single useMemo to compute both currentFontStyles and combinedStyle
    const { currentFontStyles, combinedStyle } = useMemo(() => {
        const mapperToFontFamily = mapToFontFamily ?? mapToEmbeddedFontFamily;
        const mapperToLineHeight = mapToLineHeight ?? (() => undefined);
        const currentFontStyles: FontStyleContext = {
            fontFamily: flattenedStyle.fontFamily ?? parentStyles.fontFamily,
            fontStyle: flattenedStyle.fontStyle ?? parentStyles.fontStyle,
            fontWeight: flattenedStyle.fontWeight ?? parentStyles.fontWeight,
            mapToFontFamily: mapperToFontFamily,
            mapToLineHeight: mapperToLineHeight
        };

        const mappedFontFamily = mapperToFontFamily(
            currentFontStyles.fontFamily,
            currentFontStyles.fontWeight,
            currentFontStyles.fontStyle,
        );

        const mappedLineHeight = flattenedStyle.lineHeight ?? mapperToLineHeight(flattenedStyle.fontSize);

        const combinedStyle: TextStyle = {
            ...flattenedStyle,
            fontFamily: mappedFontFamily ?? currentFontStyles.fontFamily, // Fallback to original fontFamily if mapping fails
            fontStyle: mappedFontFamily ? undefined : currentFontStyles.fontStyle, // Fallback to original fontStyle if mapping fails
            fontWeight: mappedFontFamily ? undefined : currentFontStyles.fontWeight, // Fallback to original fontWeight if mapping fails
            lineHeight: mappedLineHeight,
        };

        return { currentFontStyles, combinedStyle };
    }, [flattenedStyle, parentStyles]);

    return (
        <MyTextContext.Provider value={currentFontStyles}>
            <Text ref={ref} style={combinedStyle} {...props}>
                {children}
            </Text>
        </MyTextContext.Provider>
    );
});
