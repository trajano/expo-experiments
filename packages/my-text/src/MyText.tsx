import { createContext, forwardRef, useContext } from 'react';
import { Text, type TextProps } from 'react-native';
import {
  mapToEmbeddedFontFamily,
  mapToGoogleFontKey,
} from './mapToEmbeddedFontFamily';
import type {
  FontFamilyMapper,
  FontStyleContext,
  LineHeightMapper,
} from './MyText.types';
import { useFontStyles } from './useFontStyles';

const defaultMapToLineHeight: LineHeightMapper = () => undefined;
/**
 * MyTextContext is a React Context that provides the font styles to
 * all nested MyText components.
 */
const MyTextContext = createContext<FontStyleContext>({
  mapToFontFamily: mapToGoogleFontKey,
  mapToLineHeight: defaultMapToLineHeight,
});

export type MyTextProps = TextProps & {
  /**
   * An optional mapper that would get the final font family given the logical font family, font weight, and font style.
   * This defaults to {@link mapToGoogleFontKey}
   */
  mapToFontFamily?: FontFamilyMapper;

  /**
   * An optional mapper that would take the fontSize and translate it to a lineHeight value to ensure consistency across platforms.
   * This defaults to always return `undefined` which lets React Native handle it natively.
   */
  mapToLineHeight?: LineHeightMapper;
};

/**
 * A factory function that creates a custom Text component that inherits font styles
 * (fontFamily, fontStyle, fontWeight) from its parent MyText components unless explicitly overridden.
 *
 * @param defaultMapToFontFamily - The default function to map the font family.
 * @returns A React component that renders a Text element with inherited font styles.
 */
const createMyTextComponent = (defaultMapToFontFamily: FontFamilyMapper) =>
  // eslint-disable-next-line react/display-name
  forwardRef<Text, MyTextProps>(
    ({ children, mapToFontFamily, mapToLineHeight, style, ...props }, ref) => {
      const parentStyles = useContext(MyTextContext);
      const { currentFontStyles, combinedStyle } = useFontStyles(
        mapToFontFamily ?? defaultMapToFontFamily,
        mapToLineHeight ?? defaultMapToLineHeight,
        style,
        parentStyles,
      );

      return (
        <MyTextContext.Provider value={currentFontStyles}>
          <Text ref={ref} style={combinedStyle} {...props}>
            {children}
          </Text>
        </MyTextContext.Provider>
      );
    },
  );

/**
 * MyText is a custom Text component that inherits font styles
 * (fontFamily, fontStyle, fontWeight) from its parent MyText components
 * unless explicitly overridden. It presumes the use of Expo Google Fonts
 * and the fonts are loaded using the `useExpoGoogleFonts` hook.
 */
export const MyText = createMyTextComponent(mapToGoogleFontKey);

/**
 * MyTextE is a custom Text component similar to MyText, but it uses
 * embedded font families instead of Google Fonts by default.
 */
export const MyTextE = createMyTextComponent(mapToEmbeddedFontFamily);
