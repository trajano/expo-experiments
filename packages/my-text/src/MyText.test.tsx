import { render } from '@testing-library/react-native';
import { MyText } from './MyText'; // Adjust the import path as necessary
describe('MyText component', () => {
  it('should apply the correct Google Font key based on fontFamily, fontWeight, and fontStyle', () => {
    const { getByText } = render(
      <MyText
        style={{
          fontFamily: 'Nunito',
          fontWeight: 'bold',
          fontStyle: 'italic',
        }}
      >
        Test Text
      </MyText>,
    );

    const textElement = getByText('Test Text');

    expect(textElement.props.style.fontFamily).toBe('Nunito_700Bold_Italic');
    expect(textElement.props.style.fontStyle).toBeUndefined();
    expect(textElement.props.style.fontWeight).toBeUndefined();
  });

  it('should fallback to the original values if Google Font key mapping fails', () => {
    const { getByText } = render(
      <MyText style={{ fontWeight: '500', fontStyle: 'italic' }}>
        Test Text
      </MyText>,
    );

    const textElement = getByText('Test Text');

    expect(textElement.props.style.fontFamily).toBeFalsy();
    expect(textElement.props.style.fontStyle).toBe('italic');
    expect(textElement.props.style.fontWeight).toBe('500');
  });

  it('should inherit font styles from the parent MyText component', () => {
    const { getByText } = render(
      <MyText style={{ fontFamily: 'Nunito', fontWeight: '300' }}>
        <MyText>Child Text</MyText>
      </MyText>,
    );

    const textElement = getByText('Child Text');

    expect(textElement.props.style.fontFamily).toBe('Nunito_300Light');
    expect(textElement.props.style.fontStyle).toBeUndefined();
    expect(textElement.props.style.fontWeight).toBeUndefined();
  });

  it('should override inherited font styles if explicitly provided', () => {
    const { getByText } = render(
      <MyText style={{ fontFamily: 'Nunito', fontWeight: '300' }}>
        <MyText style={{ fontWeight: 'bold' }}>Child Text</MyText>
      </MyText>,
    );

    const textElement = getByText('Child Text');

    expect(textElement.props.style.fontFamily).toBe('Nunito_700Bold');
    expect(textElement.props.style.fontStyle).toBeUndefined();
    expect(textElement.props.style.fontWeight).toBeUndefined();
  });

  it('should override inherited font styles if explicitly provided to remove italic', () => {
    const { getByText } = render(
      <MyText style={{ fontFamily: 'Nunito', fontStyle: 'italic' }}>
        <MyText style={{ fontStyle: 'normal' }}>Child Text</MyText>
      </MyText>,
    );

    const textElement = getByText('Child Text');

    expect(textElement.props.style.fontFamily).toBe('Nunito_400Regular'); // Expect 'Nunito_400Regular' because no fontWeight was specified
    expect(textElement.props.style.fontStyle).toBeUndefined(); // fontStyle should be undefined since it's handled by the fontFamily
    expect(textElement.props.style.fontWeight).toBeUndefined(); // fontWeight should also be undefined
  });
  it('should preserve the fontVariant small-caps style', () => {
    const { getByText } = render(
      <MyText
        style={{ fontFamily: 'SourceSansPro', fontVariant: ['small-caps'] }}
      >
        Small Caps Text
      </MyText>,
    );

    const textElement = getByText('Small Caps Text');

    expect(textElement.props.style.fontVariant).toContain('small-caps');
    expect(textElement.props.style.fontFamily).toBe('SourceSansPro_400Regular');
    expect(textElement.props.style.fontWeight).toBeUndefined(); // fontWeight should be undefined since it's handled by the fontFamily
    expect(textElement.props.style.fontStyle).toBeUndefined(); // fontStyle should be undefined since it's handled by the fontFamily
  });

  it('should preserve the correct styles across three levels of nesting with SourceSansPro, even with random inner styles', () => {
    const { getByText } = render(
      <MyText
        style={{
          fontFamily: 'SourceSansPro',
          fontWeight: '600',
          fontStyle: 'italic',
          fontVariant: ['small-caps'],
        }}
      >
        <MyText style={{ fontStyle: 'normal' }}>
          <MyText style={{ fontWeight: 'bold' }}>Nested Text</MyText>
        </MyText>
      </MyText>,
    );

    const textElement = getByText('Nested Text');

    expect(textElement.props.style.fontVariant).toBeUndefined(); // it should be undefined because font variant isn't being managed.
    expect(textElement.props.style.fontFamily).toBe('SourceSansPro_700Bold');
    expect(textElement.props.style.fontWeight).toBeUndefined(); // fontWeight should be undefined since it's handled by the fontFamily
    expect(textElement.props.style.fontStyle).toBeUndefined(); // fontStyle should be undefined since it's handled by the fontFamily
  });

  it('should preserve the correct styles specifically the variants across three levels of nesting with SourceSansPro, even with random inner styles', () => {
    const { getByText } = render(
      <MyText
        style={{
          fontFamily: 'SourceSansPro',
          fontWeight: '600',
          fontStyle: 'italic',
          fontVariant: ['small-caps'],
        }}
      >
        <MyText style={{ fontStyle: 'normal' }}>
          <MyText
            style={{ fontWeight: 'bold', fontVariant: ['common-ligatures'] }}
          >
            Nested Text
          </MyText>
        </MyText>
      </MyText>,
    );

    const textElement = getByText('Nested Text');

    expect(textElement.props.style.fontVariant).toContain('common-ligatures');
    expect(textElement.props.style.fontFamily).toBe('SourceSansPro_700Bold');
    expect(textElement.props.style.fontWeight).toBeUndefined(); // fontWeight should be undefined since it's handled by the fontFamily
    expect(textElement.props.style.fontStyle).toBeUndefined(); // fontStyle should be undefined since it's handled by the fontFamily
  });

  it('should correctly apply styles to each text segment with nested MyText elements and different font families', () => {
    const { getByText, toJSON } = render(
      <MyText
        style={{
          fontFamily: 'SourceSansPro',
          fontWeight: '600',
          fontStyle: 'italic',
          fontVariant: ['small-caps'],
        }}
      >
        foo
        <MyText style={{ fontFamily: 'Nunito', fontStyle: 'normal' }}>
          bar
          <MyText style={{ fontFamily: 'SourceSansPro', fontWeight: 'bold' }}>
            baz
          </MyText>
          mew
        </MyText>
      </MyText>,
    );

    // Validate styles for the full text 'foobarbazmew'
    const foobarbazmew = getByText('foobarbazmew');

    // Validate styles for 'foo'
    expect(foobarbazmew.props.style.fontVariant).toContain('small-caps');
    expect(foobarbazmew.props.style.fontFamily).toBe(
      'SourceSansPro_600SemiBold_Italic',
    );
    expect(foobarbazmew.props.style.fontWeight).toBeUndefined();
    expect(foobarbazmew.props.style.fontStyle).toBeUndefined();

    const barbazmew = getByText('barbazmew');
    // Validate styles for 'bar'
    expect(barbazmew.props.style.fontVariant).toBeUndefined(); // No small-caps in 'bar' text
    expect(barbazmew.props.style.fontFamily).toBe('Nunito_600SemiBold');
    expect(barbazmew.props.style.fontWeight).toBeUndefined();
    expect(barbazmew.props.style.fontStyle).toBeUndefined();

    const baz = getByText('baz');
    // Validate styles for 'baz'
    expect(baz.props.style.fontVariant).toBeUndefined(); // No small-caps in 'baz' text
    expect(baz.props.style.fontFamily).toBe('SourceSansPro_700Bold');
    expect(baz.props.style.fontWeight).toBeUndefined();
    expect(baz.props.style.fontStyle).toBeUndefined();

    expect(toJSON()).toMatchSnapshot();

  });
});
