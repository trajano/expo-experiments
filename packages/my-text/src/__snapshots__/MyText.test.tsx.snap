// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyText component should correctly apply styles to each text segment with nested MyText elements and different font families 1`] = `
<Text
  style={
    {
      "fontFamily": "SourceSansPro_600SemiBold_Italic",
      "fontStyle": undefined,
      "fontVariant": [
        "small-caps",
      ],
      "fontWeight": undefined,
      "lineHeight": undefined,
    }
  }
>
  foo
  <Text
    style={
      {
        "fontFamily": "Nunito_600SemiBold",
        "fontStyle": undefined,
        "fontWeight": undefined,
        "lineHeight": undefined,
      }
    }
  >
    bar
    <Text
      style={
        {
          "fontFamily": "SourceSansPro_700Bold",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "lineHeight": undefined,
        }
      }
    >
      baz
    </Text>
    mew
  </Text>
</Text>
`;
